vectorstores/alloydb/vectorestores.go

```diff
@@ -0,0 +1,346 @@
+		}
+		results = append(results, resultMap)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("rows iteration error: %w", err)
+	}
+	return results, nil
+}
+
+func (vs *VectorStore) processResultsToDocuments(results []map[string]any) ([]schema.Document, error) {
+	var documents []schema.Document
+	for _, row := range results {
+		metadata := make(map[string]any)
+		if vs.metadataJsonColumn != "" && row[vs.metadataJsonColumn] != nil {
+			if jsonBytes, ok := row[vs.metadataJsonColumn].([]byte); ok {
+				if err := json.Unmarshal(jsonBytes, &metadata); err != nil {
+					return nil, fmt.Errorf("failed to unmarshal metadata JSON: %w", err)
+				}
+			} else {
+				return nil, fmt.Errorf("expected byte slice for metadata JSON, but got %T", row[vs.metadataJsonColumn])
+			}
+		}
+		for _, col := range vs.metadataColumns {
+			if val, ok := row[col]; ok {
+				metadata[col] = val
+			}
+		}
+		document := schema.Document{
+			PageContent: row[vs.contentColumn].(string),
+			Metadata:    metadata,
+		}
+		distance, ok := row["distance"].(float64)
+		if !ok {
+			return nil, fmt.Errorf("expected distance to be a floating value, but got %T", row["distance"])
+		}
+		document.Score = float32(distance)
+		documents = append(documents, document)
+	}
+	return documents, nil
+}
+
+// ApplyVectorIndex creates an index in the table of the embeddings
+func (vs *VectorStore) ApplyVectorIndex(ctx context.Context, index BaseIndex, name, scannIndexFunction string, concurrently bool) error {
+	if index.indexType == "exactnearestneighbor" {
+		return vs.DropVectorIndex(ctx, name)
+	}
+	function := index.distanceStrategy.searchFunction()
+	if index.indexType == "ScaNN" {
+		_, err := vs.engine.Pool.Exec(ctx, "CREATE EXTENSION IF NOT EXISTS alloydb_scann")
+		if err != nil {
+			return fmt.Errorf("failed to create alloydb scann extension: %w", err)
+		}
+	}
+	filter := ""
+	if len(index.partialIndexes) > 0 {
+		filter = fmt.Sprintf("WHERE %s", index.partialIndexes)
+	}
+	params := fmt.Sprintf("WITH %s", index.indexOptions())
+
+	if name == "" {
+		if index.name == "" {
+			index.name = vs.tableName + defaultIndexNameSuffix
+		}
+		name = index.name
+	}
+
+	concurrentlyStr := ""
+	if concurrently {
+		concurrentlyStr = "CONCURRENTLY"
+	}
+
+	stmt := fmt.Sprintf("CREATE INDEX %s %s ON %s.%s USING %s (%s %s) %s %s",
+		concurrentlyStr, name, vs.schemaName, vs.tableName, index.indexType, vs.embeddingColumn, function, params, filter)
+
+	_, err := vs.engine.Pool.Exec(ctx, stmt)
+	if err != nil {
+		return fmt.Errorf("failed to execute creation of index: %w", err)
+	}
+	_, err = vs.engine.Pool.Exec(ctx, "COMMIT")
+	if err != nil {
+		return fmt.Errorf("failed to commit index: %w", err)
+	}
+	return nil
+}
+
+// ReIndex re-indexes the VectorStore.
+func (vs *VectorStore) ReIndex(ctx context.Context, indexName string) error {
+	if indexName == "" {
+		indexName = vs.tableName + defaultIndexNameSuffix
+	}
+	query := fmt.Sprintf("REINDEX INDEX %s;", indexName)
```

Comment:
Should re-index have an option to work concurrently?

Reply:
on the python version its not concurrent! 
async def areindex(self, index_name: Optional[str] = None) -> None:
        """Re-index the vector store table."""
        index_name = index_name or self.table_name + DEFAULT_INDEX_NAME_SUFFIX
        query = f"REINDEX INDEX {index_name};"
        async with self.engine.connect() as conn:
            await conn.execute(text(query))
            await conn.commit()